### PR TITLE
[console] enhance detection and elimination of ANSI sequences

### PR DESCRIPTION
**Pull request by picnixz**
While I was implementing the line matcher object for testing Sphinx's output, I needed some 'cleaning' step based on colors. So I used `sphinx.util.console.blue` for instance; however since the function is dynamically created, it is not picked up by mypy. 

So, I added the dynamic definitions of the functions in a `TYPE_CHECKING` block. Now, I also needed some better regex for stripping colors because if we were to release that line matcher, then we don't really know whether the output could also contain other control sequences (that are currently not picked up by our regex).

In addition, we never had any test for ANSI esc. seq. so I added some. I also think that people would be interested in cleaning up other ANSI esc. seq. that could be there (not just those related to colors) but maybe they want to only remove colors and keep non-colors ones (e.g., those supported by vt100 terminals or xterms). I know the list is not complete (per https://en.wikipedia.org/wiki/ANSI_escape_code) but I don't know if I should add more (it is very important that the regex is not too generic otherwise you end up removing possible legit text).

### Remarks

- I explicitly disabled the ruff formatter for some portion of the code. The rationale behind is that I clearly don't want to have empty lines due to definitions inside `TYPE_CHECKING` blocks and I wanted to group them by 'action' (with the formatter, every function would have the same number amounts of empty lines inbetween and I didn't want to increase the size of the file's top section).
- Thanks to those new definitions, we can also remove some `# type: ignore` from mypy (which is a good thing).
- This PR should be merged before the line matcher feature since the latter is based on that one.

### CODE REVIEW

**Comment by chrisjsewell** [comment]
> So, I added the dynamic definitions of the functions in a `TYPE_CHECKING` block.

this is definitely good, perhaps though move this to a separate PR, because it feels like you are trying to do multiple things at once here?

> Now, I also needed some better regex for stripping colors

Related to #12134, I feel this is somewhat treating the symptom and not the cause; what people actually want to do, is just to run the test without colored output in the first place, not retroactively have complex regexes to remove them

**Comment by picnixz** [comment]
> this is definitely good thanks, perhaps though move this to a separate PR, because it feels like you are trying to do multiple things at once here?

Actually, if I don't add them, I cannot mypy the tests. I can separate the PR into two but then it feels weird to me to just re-add the test file.

> Related to https://github.com/sphinx-doc/sphinx/pull/12134, I feel this is somewhat treating the symptom and not the cause; what people actually want to do, is just to run the test without colored output in the first place, not retroactively have complex regexes to remove them

Yes, but since I cannot know which ones can be part of Sphinx output (if you are considering extensions), I think it's better not cleanup everything. Not that the next PR I'm preparing is (by default) removing colors.

**Comment by chrisjsewell** [comment]
> Actually, if I don't add them, I cannot mypy the tests. I can separate the PR into two but then it feels weird to me to just re-add the test file.

I mean have a PR, that would be merged before this one, that only adds:

```python
    # fmt: off
    def reset(text: str) -> str: ...  # NoQA: E704
    def bold(text: str) -> str: ...  # NoQA: E704
    def faint(text: str) -> str: ...  # NoQA: E704
    def standout(text: str) -> str: ...  # NoQA: E704
    def underline(text: str) -> str: ...  # NoQA: E704
    def blink(text: str) -> str: ...  # NoQA: E704

    def black(text: str) -> str: ...  # NoQA: E704
    def white(text: str) -> str: ...  # NoQA: E704
    def red(text: str) -> str: ...  # NoQA: E704
    def green(text: str) -> str: ...  # NoQA: E704
    def yellow(text: str) -> str: ...  # NoQA: E704
    def blue(text: str) -> str: ...  # NoQA: E704
    def fuchsia(text: str) -> str: ...  # NoQA: E704
    def teal(text: str) -> str: ...  # NoQA: E704

    def darkgray(text: str) -> str: ...  # NoQA: E704
    def lightgray(text: str) -> str: ...  # NoQA: E704
    def darkred(text: str) -> str: ...  # NoQA: E704
    def darkgreen(text: str) -> str: ...  # NoQA: E704
    def brown(text: str) -> str: ...  # NoQA: E704
    def darkblue(text: str) -> str: ...  # NoQA: E704
    def purple(text: str) -> str: ...  # NoQA: E704
    def turquoise(text: str) -> str: ...  # NoQA: E704
    # fmt: on
```

and nothing else, except removal of `type: ignore` s

**Comment by chrisjsewell** [comment]
> but since I cannot know which ones can be part of Sphinx output (if you are considering extensions)

I'm a bit confused by what you mean here, extensions use exactly the same mechanism for adding color as sphinx (at least they should be), e.g. `from sphinx.util.console import red, ...`
If you turn off color logging for sphinx, you turn off color logging for extensions

**Comment by picnixz** [comment]
>  If you turn off color logging for sphinx, you turn off color logging for extensions

Well... that's assuming they don't add filters or handlers to the logging :/. What I mean is that I cannot say *for sure* that the output is the one that *we* have constructed. But maybe I am too biased against malicious adversaries !

**Comment by picnixz** [comment]
> I mean have a PR, that would be merged before this one, that only adds:

Oh yes, that's right I did not consider that order :laughing:

**Comment by picnixz** [comment]
OK, the regex actually had issues. I think I'll go with what Chris suggested, namely, I will keep the current behaviour and only add tests.

By the way, I'm not sure if people want `strip_colors` or `strip_escape_sequences`. In general, if you have escape sequences, you just want to get rid of *all* of them, including others such as `\x1b[2K`. 

Also, I observed that our function does not work if you have something like:

> 'a' + blue('b') + '\x1b[2K' + 'c' + blue('d')

The `strip_colors` ends up removing '\x1b[2Kc' all together so... I think we should rather fix the current regex.

**Review by chrisjsewell — APPROVED** [review]
all good thanks

**Comment by picnixz** [comment]
So, what should I do with the bad strip_colors? should I just write a comment for the doc or.. ? (stripping everything appears to be fine but the steip_colors function is a bit.. wrong).

**Comment by chrisjsewell** [comment]
> So, what should I do with the bad strip_colors? should I just write a comment for the doc or.. ? (stripping everything appears to be fine but the steip_colors function is a bit.. wrong).

yeh I think that sounds a good idea 👍 

Could maybe mention that its use case is more envisaged for testing purposes (e.g. for checking warning strings), rather than any "production" code. Or at least that's my thinking on it

**Comment by picnixz** [comment]
@chrisjsewell I know you accepted the PR, but here is my final solution (together with the caution comment).

I think the tests capture most of the non-pathological cases (and I easily fixed the previous bug). By the way, to avoid namespace pollution, I renamed some module loop invariants that could shadow names if incorrectly imported.

File: sphinx/util/console.py [hunk 1]
```diff
@@ -128,11 +134,39 @@ def escseq(name: str) -> str:
...
 
 def strip_colors(s: str) -> str:
+    """Remove the ANSI color codes in a string *s*.
+
+    .. caution::
+
+       This function is not meant to be used in production and should only
+       be used for testing Sphinx's output messages.
+
+    .. seealso:: :func:`strip_control_sequences`
```
Comment (picnixz) [inline sphinx/util/console.py]:
```suggestion
    .. seealso:: :func:`strip_escape_sequences`
```

**Review by chrisjsewell — APPROVED** [review]
Yeh I think still all good thanks.

I guess making `strip_escape_sequences` public is a thinking point,
but it looks like you have created a good range of tests for it, so that should be fine 👍

**Comment by picnixz** [comment]
> I guess making strip_escape_sequences public is a thinking point,

That's why I also marked the feature as *experimental* (at least, if it doesn't work I can hide behind that justification :))

**Pull request merged**